### PR TITLE
Tighten up write permissions in Docker image

### PR DESCRIPTION
**Pull request by pugnascotia**
Recursively remove write access from the `bin`, `jdk`, `lib` and
`modules` directories, since this access is not required, and removing
it makes it harder to exploit other issues in an ES distribution.

### CODE REVIEW

**Review by droberts195 — APPROVED** [review]
LGTM assuming it doesn't cause any failures.

If it turns out something legitimately needed write access to one of these files or directories then we should look to make the minimum necessary subset of files or directories writable again.

**Comment by rjernst** [comment]
Instead of giving the entire ES install permissions and then restricting a fixed set from write permissions, can we instead explicitly give write permissions to the dirs that actually need it? Otherwise any additions (or even renames) of these directories in the future would need corresponding changes here which would be easy to miss.

**Comment by mark-vieira** [comment]
> Instead of giving the entire ES install permissions and then restricting a fixed set from write permissions, can we instead explicitly give write permissions to the dirs that actually need it?

This seems reasonable. How difficult would this be?

**Comment by pugnascotia** [comment]
I pushed a change so that all permissions are explicitly set under `/usr/share/elasticsearch`, is that what you had in mind?

**Comment by rjernst** [comment]
I was thinking more about keeping the original permissions, and just changing the owner so it is root like rpm/deb have.

**Comment by pugnascotia** [comment]
@rjernst see what you think of the latest changes.

File: distribution/docker/src/docker/Dockerfile [hunk 1]
```diff
@@ -330,10 +330,15 @@ COPY ${bin_dir}/docker-entrypoint.sh /usr/local/bin/docker-entrypoint.sh
...
+# 5. Tighten up permissions on the ES home dir (the permissions of the contents are handled earlier)
+# 6. Make it possible for a user in the `elasticsearch` group to write to certain directories
 RUN chmod g=u /etc/passwd && \\
-    chmod 0775 /usr/local/bin/docker-entrypoint.sh && \\
+    chmod 0555 /usr/local/bin/docker-entrypoint.sh && \\
     find / -xdev -perm -4000 -exec chmod ug-s {} + && \\
-    ln -sf /etc/pki/ca-trust/extracted/java/cacerts /usr/share/elasticsearch/jdk/lib/security/cacerts
+    ln -sf /etc/pki/ca-trust/extracted/java/cacerts /usr/share/elasticsearch/jdk/lib/security/cacerts && \\
+    chmod 0755 /usr/share/elasticsearch && \\
+    chgrp 1000 bin && \\
```
Comment (rjernst) [inline distribution/docker/src/docker/Dockerfile:340]: Why does bin need to be group owned by the running user? The files all are o+rx right?

File: distribution/docker/src/docker/Dockerfile [hunk 2]
```diff
@@ -330,10 +330,15 @@ COPY ${bin_dir}/docker-entrypoint.sh /usr/local/bin/docker-entrypoint.sh
...
+    chgrp -R 1000 config data logs plugins
```
Comment (rjernst) [inline distribution/docker/src/docker/Dockerfile:341]: Why does the plugins dir need to be grouped for the running user?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/FileMatcher.java [hunk 3]
```diff
@@ -37,14 +37,16 @@
         Directory
     }
 
-    public static final Set<PosixFilePermission> p775 = fromString("rwxrwxr-x");
-    public static final Set<PosixFilePermission> p770 = fromString("rwxrwx---");
-    public static final Set<PosixFilePermission> p755 = fromString("rwxr-xr-x");
-    public static final Set<PosixFilePermission> p750 = fromString("rwxr-x---");
-    public static final Set<PosixFilePermission> p660 = fromString("rw-rw----");
+    public static final Set<PosixFilePermission> p444 = fromString("r--r--r--");
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/FileMatcher.java:40]: This is now unused right?

**Review by rjernst — COMMENTED** [review]
Thanks @pugnascotia. I have a few questions.

**Comment by pugnascotia** [comment]
I realised that I had changed the image to use a default group of `elasticsearch`, but that would break the current guidance to use GID 0 for bind-mounted volumes (which e.g. OpenShift does), and probably be considered a breaking change.

Instead, the image still uses group `root` but now also uses user `root` as well. Since the image runs as `elasticsearch`, that should make it more difficult to trample files via an exploit, in conjunction with the tightened permissions.

**Comment by pugnascotia** [comment]
I went back over these changes again, and basically tried to lock things down as much as possible, while making it possible to install plugins without jumping through ridiculous hoops. See what you think.

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
@elasticmachine update branch

File: distribution/docker/src/docker/Dockerfile [hunk 4]
```diff
@@ -230,26 +230,24 @@ RUN tar -zxf /opt/elasticsearch.tar.gz --strip-components=1
...
     mv config/log4j2.properties config/log4j2.file.properties && \\
     mv config/log4j2.docker.properties config/log4j2.properties && \\
-    chmod 0660 config/elasticsearch.yml config/log4j2*.properties && \\
-    find ./jdk -type d -exec chmod 0755 {} + && \\
-    find . -xdev -perm -4000 -exec chmod ug-s {} + && \\
-    find . -type f -exec chmod o+r {} +
+    find . -type d -exec chmod 0555 {} + && \\
+    find . -type f -exec chmod 0444 {} + && \\
+    chmod 0555 bin/* jdk/bin/* jdk/lib/jspawnhelper modules/x-pack-ml/platform/linux-\$(arch)/bin/* && \\
+    chmod 0775 config config/jvm.options.d data logs plugins && \\
```
Comment (rjernst) [inline distribution/docker/src/docker/Dockerfile:249]: Including plugins here seems problematic. They are code, and should therefore not be writeable by the user that runs elasticsearch.

File: distribution/docker/src/docker/Dockerfile [hunk 5]
```diff
@@ -230,26 +230,24 @@ RUN tar -zxf /opt/elasticsearch.tar.gz --strip-components=1
...
+    mkdir data && \\
     mv config/log4j2.properties config/log4j2.file.properties && \\
     mv config/log4j2.docker.properties config/log4j2.properties && \\
-    chmod 0660 config/elasticsearch.yml config/log4j2*.properties && \\
-    find ./jdk -type d -exec chmod 0755 {} + && \\
-    find . -xdev -perm -4000 -exec chmod ug-s {} + && \\
-    find . -type f -exec chmod o+r {} +
+    find . -type d -exec chmod 0555 {} + && \\
+    find . -type f -exec chmod 0444 {} + && \\
+    chmod 0555 bin/* jdk/bin/* jdk/lib/jspawnhelper modules/x-pack-ml/platform/linux-\$(arch)/bin/* && \\
```
Comment (rjernst) [inline distribution/docker/src/docker/Dockerfile:248]: The correct files already are executable in the original archive that is extracted. Why was the executable bit removed in the first place? Seems like we should just keep it as it was, instead of trying to reform what is executable here.

File: distribution/docker/src/docker/Dockerfile [hunk 6]
```diff
@@ -287,8 +285,7 @@ RUN ${package_manager} update --setopt=tsflags=nodocs -y && \\
 
 RUN groupadd -g 1000 elasticsearch && \\
     adduser -u 1000 -g 1000 -G 0 -d /usr/share/elasticsearch elasticsearch && \\
-    chmod 0775 /usr/share/elasticsearch && \\
-    chown -R 1000:0 /usr/share/elasticsearch
+    chown -R 0:0 /usr/share/elasticsearch
```
Comment (rjernst) [inline distribution/docker/src/docker/Dockerfile:288]: The rpm/deb packages setup ownership so that the group is the elasticsearch user, which is readable but not writable. Can we just keep the original permissions defined in the distribution and fix the ownership?

File: qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java [hunk 7]
```diff
@@ -147,7 +147,7 @@ public void test041AmazonCaCertsAreInTheKeystore() {
     public void test042KeystorePermissionsAreCorrect() throws Exception {
         waitForElasticsearch(installation);
 
-        assertPermissionsAndOwnership(installation.config("elasticsearch.keystore"), p660);
+        assertPermissionsAndOwnership(installation.config("elasticsearch.keystore"), "elasticsearch", "root", p660);
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java:150]: This seems backwards. Shouldn't the group be elasticsearch and the owner be root?

**Review by rjernst — COMMENTED** [review]
I have a few more questions, and I'm not sure all my previous comments were addressed.

File: distribution/docker/src/docker/Dockerfile (hunk 4 above)
Comment (mieciu) [reply distribution/docker/src/docker/Dockerfile:249]: I remember we discussed the plugins a while ago in https://github.com/elastic/elasticsearch/issues/69533. Being able to install a plugin while running Elasticsearch as a non-root user (ideally without the need to `--group-add 0`) is vital for Cloud deployments.

**Comment by pugnascotia** [comment]
I've picked up this work again, and I've been scratching my head over what we're trying to do and what are the constraints. I think it would helpful to lay these out:

   1. Docker images are immutable, therefore we don't need ES files to be writable e.g. for an upgrade. On the contrary, most files that we ship in the image should never be overwritten.
   2. On a multi-user system, it is appropriate to install ES as `root:elasticsearch`. However in the Docker world, it is essentially a single-user system - although popular images like `postgres` and `redis` run as `root` by default and then drop to another user, because our `7.x` images also do this they are routinely flagged in security scans
   3. We recommend using GID 0 for sharing files via bind-mounts (OpenShift also recommends this, for example). Using an ownership of `root:elasticsearch` in the container would complicates this, versus `root:elasticsearch`

To sum up, I'm don't believe reusing the archive permissions is desirable. 

What I've tried to do here is use `root:root` user for immutable files and directories, and use `elasticsearch:root` when they are writable. Rather than trying to tweak file permissions here and there, I decided that it was clearer to simply reset them all to known values that are more appropriate for the Docker environment.

Plugins complicate the picture, because not only does the executing user need write access to `plugins`, but it also (technically) needs write access to `bin` because plugins can install CLI tools. Essentially, the line between installing ES and running ES becomes non-existent, so we don't install as `root` (for example) and run as `elasticsearch`.

File: qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java (hunk 7 above)
Comment (pugnascotia) [reply qa/os/src/test/java/org/elasticsearch/packaging/test/DockerTests.java:152]: The startup process creates this if necessary at startup, so it has to be `elasticsearch:root`.

**Comment by pugnascotia** [comment]
Backported to `7.x` in c04dc4ccefe.

**Comment by mark-vieira** [comment]
FYI, I've reverted the backport to `7.x` in 8effc56cfc92dffdb3b9bb900427efb1858d269e. It was causing some packaging test failures: https://gradle-enterprise.elastic.co/s/kvmrl5ln52pv6

**Comment by pugnascotia** [comment]
Bah, sorry about that. New backport in #76006. FWIW, I did run the `DockerTests` before pushing, but I forgot about the `KeystoreManagementTests`.
